### PR TITLE
make classifier thread safe

### LINKED ISSUE #112: Drivers should be thread safe.
for concurrency of machine learning, Jubatus server holds giant lock for each model.
But giant lock is harmful for performance on multi-core processors.

Drivers should be thread safe, and server should not hold giant lock for each model.

### PR DESCRIPTION
**Pull request by kumagi**
fix #112

### CODE REVIEW

**Comment by kmaehashi** [comment]
@kumagi Please remove unittset_gtest.py.

**Comment by gwtnb** [comment]
cpplint

File: jubatus/core/driver/test_util.hpp [hunk 1]
```diff
@@ -51,6 +53,21 @@ jubatus::util::lang::shared_ptr<jubatus::core::fv_converter::datum_to_fv_convert
...
 }
 
+jubatus::core::fv_converter::datum generate_random_datum(
+    jubatus::util::math::random::mtrand& rand,
+    int datum_length) {
+  jubatus::core::fv_converter::datum d;
+  std::string value;
+  value.resize(datum_length * 2 + 1);
+  for (int j = 0; j < datum_length; ++j) {
+    value[j * 2]     = (rand() % ('z' - 'a')) + 'a';
```
Comment (gwtnb) [inline jubatus/core/driver/test_util.hpp]: 'z' - 'a' is 25 if a-z are sequential.

File: jubatus/core/driver/test_util.hpp [hunk 2]
```diff
@@ -51,6 +53,21 @@ jubatus::util::lang::shared_ptr<jubatus::core::fv_converter::datum_to_fv_convert
   return converter;
 }
 
+jubatus::core::fv_converter::datum generate_random_datum(
+    jubatus::util::math::random::mtrand& rand,
+    int datum_length) {
+  jubatus::core::fv_converter::datum d;
+  std::string value;
+  value.resize(datum_length * 2 + 1);
```
Comment (gwtnb) [inline jubatus/core/driver/test_util.hpp]: `resize` fills `'\0'`. It is not recommended to set '\0' in std::string when the string is used as string.

File: jubatus/core/driver/test_util.hpp [hunk 3]
```diff
@@ -51,6 +53,21 @@ jubatus::util::lang::shared_ptr<jubatus::core::fv_converter::datum_to_fv_convert
...
+  for (int j = 0; j < datum_length; ++j) {
+    value[j * 2]     = (rand() % ('z' - 'a')) + 'a';
+    value[j * 2 + 1] = ' ';
+  }
+  value[datum_length*2] = '\0';
```
Comment (gwtnb) [inline jubatus/core/driver/test_util.hpp]: Assigning `'\0'`to `std::string` should be avoided when the `std::string` object is used as string.

Comment (gwtnb) [inline jubatus/core/driver/test_util.hpp]: So a possible implementation is as follows:

```
std::string value(datum_length * 2, ' ');
for (int j = 0; j < datum_length; ++j) {
  value[j*2] = (rand() % 26) + 'a'; // adding int to 'a' is not so good
}
```

**Comment by gwtnb** [comment]
cpplint.

File: jubatus/core/driver/test_util.hpp (hunk 1 above)
Comment (kumagi) [inline jubatus/core/driver/test_util.hpp]: OK, I'll make it 26.

File: jubatus/core/driver/test_util.hpp (hunk 2 above)
Comment (kumagi) [inline jubatus/core/driver/test_util.hpp]: OK, I'll eliminate `\0`.

**Comment by kmaehashi** [comment]
@kumagi I'll add Valgrind test to CI. Please paste the command line option here.

**Pull request merged**